x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java

```diff
@@ -174,29 +174,130 @@ protected void readInternal(final byte[] buffer, final int offset, final int len
         }
-        final Tuple<Long, Long> range = computeRange(IntStream.range(0, part).mapToLong(fileInfo::partBytes).sum());
-        assert assertRangeIsAlignedWithPart(range);
+        final Tuple<Long, Long> partRange = computeRange(IntStream.range(0, part).mapToLong(fileInfo::partBytes).sum());
+        assert assertRangeIsAlignedWithPart(partRange);
+
         try {
             final CacheFile cacheFile = getCacheFileSafe();
             try (ReleasableLock ignored = cacheFile.fileLock()) {
-                final int bytesRead = cacheFile.fetchRange(range.v1(), range.v2(), (start, end) -> {
-                    logger.trace("range [{}-{}] of file [{}] is now available in cache", start, end, fileInfo.physicalName());
-                    return Math.toIntExact(end - start);
-                }, (start, end) -> writeCacheFile(cacheFile.getChannel(), start, end)).get();
 
-                assert bytesRead == (range.v2() - range.v1());
-                return bytesRead;
+                final Tuple<Long, Long> range = cacheFile.getAbsentRangeWithin(partRange.v1(), partRange.v2());
+                if (range == null) {
+                    logger.trace(
+                        "prefetchPart: part [{}] bytes [{}-{}] is already fully available for cache file [{}]",
+                        part,
+                        partRange.v1(),
+                        partRange.v2(),
+                        cacheFileReference
+                    );
+                    return;
+                }
+
+                final long rangeStart = range.v1();
+                final long rangeEnd = range.v2();
+                final long rangeLength = rangeEnd - rangeStart;
+
+                logger.trace(
+                    "prefetchPart: prewarming part [{}] bytes [{}-{}] by fetching bytes [{}-{}] for cache file [{}]",
+                    part,
+                    partRange.v1(),
+                    partRange.v2(),
+                    rangeStart,
+                    rangeEnd,
+                    cacheFileReference
+                );
+
+                final FileChannel fc = cacheFile.getChannel();
+                assert assertFileChannelOpen(fc);
+                final byte[] copyBuffer = new byte[Math.toIntExact(Math.min(COPY_BUFFER_SIZE, rangeLength))];
+
+                long totalBytesRead = 0L;
+                long remaining = rangeEnd - rangeStart;
+                final long startTimeNanos = stats.currentTimeNanos();
+                try (InputStream input = openInputStream(rangeStart, rangeLength)) {
+                    while (remaining > 0L) {
+                        assert totalBytesRead + remaining == rangeLength;
+                        final int bytesRead = readSafe(input, copyBuffer, rangeStart, rangeEnd, remaining, cacheFileReference);
+                        final long readStart = rangeStart + totalBytesRead;
+                        cacheFile.fetchRange(readStart, readStart + bytesRead, (start, end) -> {
```

Comment:
Can we capture the returned Future here and assert in the next line that `future.isDone`?

Reply:
No, the future might not be done on return if another thread already acquired the target range and hasn't filled it in yet. That's ok, tho, we assume it'll succeed eventually and keep on consuming the blob until we find a range that hasn't already been acquired by anyone else.